### DISCUSSION TITLE
Dokku not build with java?

### DISCUSSION THREAD

**Discussion by ibmoreno** [dokku/dokku#5374] [General]
Sorry for my english, I'm from Brazil. I'm starting on dokku, I'm not able to build a spring-boot java application, I have experience with heroku, but I can't understand why dokku doesn't detect my project as java, does dokku support java? should i install java and maven in the environment? The articles I see with java are few and they don't do any configuration with java and they work. What is wrong?

my project

https://github.com/ibmoreno/exemplo

Erro
``` git push dokku master
Enumerating objects: 58, done.
Counting objects: 100% (58/58), done.
Delta compression using up to 4 threads
Compressing objects: 100% (44/44), done.
Writing objects: 100% (58/58), 58.86 KiB | 1.47 MiB/s, done.
Total 58 (delta 6), reused 0 (delta 0), pack-reused 0
remote:  !     Herokuish builder not supported on arm64 servers.
remote:  !     Switching to pack builder.
remote: -----> Cleaning up...
remote: -----> Building exemplo from cnb stack heroku/buildpacks (experimental)...
remote:  !     Missing pack, install it
remote:  !     Removing invalid image tag dokku/exemplo:latest
remote:  !     App build failed
To xxx.xxx.xxx.xxx:exemplo
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'xxx.xxx.xxx.xxx:exemplo'`
```
NOTA: XXX mask my number IP

**Comment by josegonzalez** [comment]
The error message shows what is going on here:

```
remote:  !     Missing pack, install it
```

My guess is you changed to the `pack` builder instead of using the default `herokuish` builder. If you switch back _or_ install pack, the build will proceed.

You can follow the instructions [here](https://buildpacks.io/docs/tools/pack/) to install it.

**Reply by ibmoreno** [reply]
Hi josegonalez, but it's standard installation, it seems that the standard doesn't support arm architecture as you can see in the previous errors. searching a lot and I see people running it without any problem, without needing to install anything else. does the default dokku installation not cover java build?

`remote:  !     Herokuish builder not supported on arm64 servers.`

thank you so much

**Reply by josegonzalez** [reply]
The error there is correct. The herokuish builder is not supported on ARM/ARM64, and there are no plans to support it. Pack _is_ supported, but you'll need to use an ARM/ARM64 builder image.

One way that should work is our Dockerfile builder, and the upcoming nixpacks builder.

**Reply by ibmoreno** [reply]
Exactly, I started using this other image but the environment variables error, with everything I ended up giving up and rethinking the platform issue, I'm using Oracle Cloud.

https://hub.docker.com/r/healthcareblocks/herokuish

dokku buildpacks:set-property example stack healthcareblocks/herokuish:latest
Mais sobre o texto original

**Reply by josegonzalez** [reply]
Ah interesting, I didn't realize there was an arm version of it. Seems _quite_ old though (2 years) and maybe unmaintained.

**Comment by josegonzalez** [comment]
Just a note: the latest herokuish release - v0.5.40 - works on arm/arm64 (though the buildpacks will likely mostly fail due to referencing amd64 files) and the upcoming Dokku 0.29.x release will support force-enabling it: https://github.com/dokku/dokku/pull/5478

**Reply by ibmoreno** [reply]
Hi josegonzalez,

Good news, I'll wait for the next version of Dokku 0.29 to test, I've been trying to use it in the oracle cloud. Thanks a lot for the news.
